jeol.py:  Add lazy loading for spectrum image, and add loader for SEM/STEM image in pts file

nem1234 opened the pull request:
### Description of the change
- Add file reader for SEM/STEM image in pts file
- Add lazy loading for spectrum image in pts file
- Add function to load selected frames in pts file
- Enable multpule image loading from pts file for asw file loader.
- Add test code and test images for incomplete frames.

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np
s = hs.load( filename.pts, si_lazy=True, read_em_image=True, frame_list=[0,2,3])
```

nem1234 committed:
Add test files

nem1234 committed:
Modify asw file reader to accept multiple image from pts reader

nem1234 committed:
bugfix, remove extra append to dictionary

nem1234 committed:
Merge branch 'hyperspy:RELEASE_next_minor' into RELEASE_next_minor

nem1234 committed:
Add support for pts file with broke Sweep data
(Another type, sweep value includes incomplete frame)

nem1234 committed:
Remove "sweep" setting after reading data

nem1234 committed:
Add document about options for pts loader

nem1234 committed:
Add test for SEM/STEM image in pts files

nem1234 committed:
remove blank at end of lines

nem1234 committed:
Add lazy data loading

nem1234 commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -94,13 +94,19 @@ def file_reader(filename, **kwds):
                                         reader_function = extension_to_reader_mapping[sub_ext]
                                         d = reader_function(file_path, scale, **kwds)
-                                        dictionary.append(d)
+                                        if isinstance(d, list):
```
PTS file loader now returns spectrum image and optionally SEM/TEM image.
So, asw file reader need to accept list of images.

nem1234 commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -267,8 +319,45 @@ def check_multiple(factor, number, string):
 
-        axes = [
+        data_shape = [height, width, channel_number]
+        # Sweep value is not reliable
```
Sweep value in the pts file header is not reliable.
In some case, sweep value shows the number of valid frames, and in another case, it shows sum of the numbers of valid and incomplete frame (may be version dependent.
So, pre-scanning the pts file is needed to check it.
(Original JEOL software also shows wrong number of frames, and operation with last incomplete frame seems unstable)

ericpre replied:
Can you please add these comment in the code too? :)
For future reference, it is useful to the explanations supporting the observations! Also, when it is written in the code, it can save us having to dig on github using `git blame` with the hope of find the reasoning.

nem1234 replied:
added comments in python code

nem1234 commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -454,63 +543,242 @@ def parsejeol(fd):
-@numba.njit(cache=True)
-def readcube_frames(rawdata, hypermap, sweep, only_valid_data, rebin_energy, channel_number,
-             width_norm, height_norm, max_value):  # pragma: no cover
+def readcube(rawdata, frame_ptr_list, frame_list, 
```
readcube and readcube_frames functions are merged into one function.
Inner loop "readframe_dense"(originally it is part of readcube) and "readframe_sparse"(for lazy loading) are optimized by numba.
Outer loop, readcube does not use numba.

nem1234 commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -454,63 +543,242 @@ def parsejeol(fd):
+    n_frames = len(frame_list)
+    EM_dtype = np.uint16
+    frame_step = 1
+    if sum_frames:
```
In case of sum_frames,  spectrum image and SEM/STEM image are summing up to same frame.
To avoid overflow on integration of SEM/STEM image, data type of np.uint32 is selected for 16 frames and over.
(range of image intensity in each frame is 0-4095 (0-0xfff))

ericpre replied:
Yes, this sounds very sensible indeed! Can you please add this comment in the code itself for future reference: at first read, the `n_frames > 16:` seems arbitrary but it is not at all!

nem1234 committed:
Revise document to re-format as style guide.
Bug fix about the data dimensions when sum_frames=True and si_lazy=True

nem1234 committed:
update comments

ericpre commented on doc/user_guide/io.rst:
```diff
@@ -1807,28 +1807,43 @@ files automatically.
+  for ``pts`` file only, frames in frame_list will be loaded.
+  for example, ``frame_list=[1,3]`` means second and forth frame is selected.
+  If ``None``, all frames are loaded.
+- ``si_lazy`` : bool, default False
```
You prefer not to rely on the standard `lazy` argument to avoid loading images lazily too?

nem1234 replied:
merge to standard "lazy" option.

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -267,8 +319,45 @@ def check_multiple(factor, number, string):
+        frame_list2 = []
+        for frame_idx in frame_list:
+            if frame_idx < 0 or frame_idx >= sweep:
+                _logger.info(f"Ignoreing frame {frame_idx} : Selected frame is not found in pts data.")
```
```suggestion
                _logger.info(f"Ignoring frame {frame_idx} : Selected frame is not found in pts data.")
```

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -267,8 +319,45 @@ def check_multiple(factor, number, string):
+        data_shape = [height, width, channel_number]
+        # Sweep value is not reliable
+        # sweep = meas_data_header["Doc"]["Sweep"]
+        frame_ptr_list, last_valid, has_em_image = pts_prescan(rawdata, width, height)
```
The drawback of this approach is to have to go through the dataset twice, which for large file is not insignificant. An alternative may be to try to use the unreliable `meas_data_header["Doc"]["Sweep"]` and if it works fine, then we continue otherwise, we fallback to the slower route of the "prescan" followed by the actual read.
I am not sure if I understand correctly the issue and therefore if this alternative is sensible... 

To make a sensible judgement call, it would good to have an idea if this issue is usual or it is only limited to some version of software, etc. In case, this issue isn't relevant for most systems, it would be shame to always double the reading time.

ericpre replied:
Add `frame_ptr_list` to the `original_metadata` dictionary, so that it could be reused when reopening the file, for example to align frames?

nem1234 committed:
Get chunks from si_lazy option

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -267,8 +319,45 @@ def check_multiple(factor, number, string):
+                _logger.info(f"Ignoreing frame {frame_idx} : Selected frame is not found in pts data.")
+                continue
+            frame_list2.append(frame_idx)
+        frame_list = np.array(frame_list2)
```
It is a bit suggestive, I find the following more clear:
```suggestion
        if not frame_list:
            frame_list = np.arange(sweep)
        frame_list = np.asarray(frame_list)
        wrong_frames_idx = np.argwhere((frame_list<0) | (frame_list>=sweep))
        if len(wrong_frames_idx) > 0:
            wrong_frames = frame_list[wrong_frames_idx].flatten().tolist()
            _logger.info(f"Ignoring frame {wrong_frames} : Selected frame is not found in pts data.")
```

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -454,63 +543,244 @@ def parsejeol(fd):
+    """
+    for value in rawdata:
+        dtype = value & 0xf000
+        dval = value & 0xfff
```
Maybe as for `pts_prescan` use `value_type` and `value` since `dtype` is used widely in numpy? The use of `dtype` is a bit misleading for the reader.

nem1234 committed:
Bugfix: summing up SEM/STEM images
Add: Summing up with shifting frames

nem1234 commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -186,7 +192,55 @@ def read_img(filename, scale=None, **kwargs):
+        (usually interrupting mesurement makes incomplete frame)
+    read_em_image : bool, default False
+        read SEM/STEM image from pts file if read_em_image == True
+    frame_list : list of int [fr0, fr1, ...] or array [[fr0, dx0,dy0,dz0],...], default None
```
add shifting of frames for express drift correction.
due to the memory usage, it is difficult to make drift correction after reading 4-dimensional data

ericpre replied:
What's about using two different arguments, for example, `frame_selection` and `drift_correction` instead of `frame_list` only? Conceptually, there are different thing and this would be easier to describe in the docstring and user guide.

Also it would be good to be able to do the following:

```python
s_list = hs.load('file.asw', select_type='images', sum_frames=False)
s_images = s_list[0] # select the correct signal
shifts = s_images.estimate_shift2D()
hs.load('file.asw', drift_correction=shifts, sum_frames=False)
```

If the `frame_ptr_list` is saved to `original_metadata`
```python
frame_ptr_list = s_images.original_metadata.frame_ptr_list
hs.load('file.asw', drift_correction=shifts, sum_frames=False, frame_start_index=frame_ptr_list)

nem1234 replied:
frame_shifts option is introduced that is compatible to the result of estimate_shift2D

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -454,63 +558,258 @@ def parsejeol(fd):
-            channel_number, width_norm, height_norm, max_value):  # pragma: no cover
-    frame_idx = 0
-    previous_y = 0
+def pts_prescan(rawdata, width, height):   # pragma: no cover
```
Is this function as slow as `readframe_dense`? If so, this is going to take a significant portion of the loading time (~50%?) and it may be worth trying to find an alternative way to know where the frames start.

nem1234 replied:
Redesign pts reader

ericpre reviewed (commented):
There is a lot of good work in this PR! I have left some questions/comments for discussion.

nem1234 committed:
Re-design one-pass PTS reader

nem1234 committed:
modify document to add new option "frame_shifts"

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -186,7 +192,54 @@ def read_img(filename, scale=None, **kwargs):
              SI_dtype=np.uint8, cutoff_at_kV=None, downsample=1,
-             only_valid_data=True, **kwargs):
+             only_valid_data=True, read_em_image=False,
+             frame_list=[], frame_start_index=[], frame_shifts=[], 
```
It is usually better not to set a default value for argument, which are of mutable type - see https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html.

A better pattern is set these values to `None` in the definition and inside the function, set the default value when it is `None`.

It is possibly not an issue here, but more a good practise to enforce!

nem1234 replied:
fixed.

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -359,18 +447,41 @@ def check_multiple(factor, number, string):
+                "original_filename": os.path.basename(filename),
+                "date": datefile.date().isoformat(),
+                "time": datefile.time().isoformat(),
+                "title": "SEM/STEM Image"
```
In case, there are several images, it is possible to retrieve some information from the header to differentiate these images and add it to the title?

nem1234 replied:
May be fixed.
new title should be   "S(T)EM Image extracted from xxxx.pts".

ericpre replied:
Sorry, my comment wasn't very clear, I was referring to the type of detector, e.g. `BF`, `DF`, etc. If this information is available, it is useful to have it in the title, for example, when one want to do some processing on images from a specific detector, etc.
We should leave it for another PR! :)

nem1234 replied:
I'm sorry, but I couldn't find the information of SEM/STEM detector/lens in the pts file header.
Also, the difference of SEM and TEM is not clear, and in the original code, the Accel Voltage is used to set the mode of SEM/TEM.

nem1234 replied:
The JEOL EDS software seems to made for SEM, and it communicate with TEM for a limited number of parameters like magnification and accel voltage. Some parameters in the header are filled with strange values, for example, the value of beam current. The image in JEOL EDS software always shows "BF"/"SEI" even if the ADF detector is used for STEM imaging.

ericpre replied:
OK, thanks for the clarification!

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -453,64 +564,306 @@ def parsejeol(fd):
+            # incomplete data, not accepted
+            if sum_frames:
+                # cancel last frame
+                print("Removing data of last frame = ", frame_idx)
```
We use the logger instead of calling `print` so that the user can control the level of verbosity.

nem1234 replied:
This is a debugging code.
already removed.

nem1234 committed:
Add comments, remove debug print

nem1234 committed:
fix bug around frame shift

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -453,64 +583,383 @@ def parsejeol(fd):
+        ptr += flen
+        frame_count += 1
+    if sum_frames:
+        data_shape = [height, width, channel_number]
+        ar_s = sparse.COO(v[1:4], v[4], shape=data_shape)
+    else:
+        data_shape = [frame_count, height, width, channel_number]
+        ar_s = sparse.COO(v[0:4], v[4], shape=data_shape)
```
Use `DenseSliceCOO` from https://github.com/hyperspy/hyperspy/blob/b5815c01d303d6598484d97096622708b38e41e0/hyperspy/misc/io/fei_stream_readers.py#L26-L35

```suggestion
        data_shape = [height, width, channel_number]
        ar_s = DenseSliceCOO(v[1:4], v[4], shape=data_shape)
    else:
        data_shape = [frame_count, height, width, channel_number]
        ar_s = DenseSliceCOO(v[0:4], v[4], shape=data_shape)
```

To avoid the following error:
```python

  File "/home/eric/Dev/hyperspy/hyperspy/_signals/lazy.py", line 127, in compute
    self.data = data

  File "/home/eric/Dev/hyperspy/hyperspy/signal.py", line 2411, in data
    self._data = np.atleast_1d(np.asanyarray(value))

  File "/opt/miniforge/lib/python3.9/site-packages/numpy/core/_asarray.py", line 171, in asanyarray
    return array(a, dtype, copy=False, order=order, subok=True)

  File "/opt/miniforge/lib/python3.9/site-packages/sparse/_sparse_array.py", line 229, in __array__
    raise RuntimeError(

RuntimeError: Cannot convert a sparse array to dense automatically. To manually densify, use the todense method.
```

nem1234 replied:
Thank you.
Fixed as your suggestion

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -453,64 +583,383 @@ def parsejeol(fd):
+            return hypermap[:target_frame_num,:height,:width], em_image[:target_frame_num,:height,:width], has_em_image, frame_num, frame_start_index, valid, max_shift
+        
+    # for lazy loading
+    length = 0
+    for d in data_list:
+        length += len(d)
```
```suggestion
    length = np.sum([len(d) for d in data_list])
```

nem1234 replied:
fixed.

ericpre commented:
@nem1234, is it ready for review? This is already quite a lot in this PR and it would be good to finish it and get it merged.
It is easier to work with small self-contained PR (to review and discuss), rather than large PR, it you want to implement other feature, it would be better to do it in a separate PR.

nem1234 committed:
commit files those are missed in the previous commit

nem1234 committed:
Remove debug print

nem1234 commented:
This PR become ready for review.
I'm sorry it become large.

Performance for lazy loading (dask - COO array) is still not good.
I'll revise it in near future.

nem1234 committed:
bugfix: invalid buffer size for non-square pts
bugfix: validity check when downsample>1

nem1234 commented:
due to my limited knowledge on git, recent codes was not pushed to here.
The performance of lazy loading is not good yet. 
I'll brush up it may be in the next PR.

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -186,7 +192,54 @@ def read_img(filename, scale=None, **kwargs):
+    	list of frame numbers to be read (None for all data)
+    	if dx, dy, dz are specified, each images are shifted. dx, dy, dz are integers, in unit of pixel.
+	This is useful for express drift correction. (not suitable for a detailed analysis)
+    si_lazy : bool or tuple of chunksize, default False
+    	read spectrum image into sparse array if si_lazy == True
+    	SEM/STEM image is always read into dense array (np.ndarray)
+    	if tuple is specified, dask.array is generated using chunksize=si_lazy
```
The si_lazy option has been removed:
```suggestion
```

nem1234 replied:
fixed.

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -186,7 +192,54 @@ def read_img(filename, scale=None, **kwargs):
+    	if tuple is specified, dask.array is generated using chunksize=si_lazy
+    lazy : bool, default False
+	set lazy flag not only spectrum image but also other data,
+	if lazy == True. This also set si_lazy = True.
+	Only the spectrum image data is read as a sparse array,
```
Update since `si_lazy` has been removed and also mention sparse array - see other comments regarding lazy loading/sparse array.

nem1234 replied:
fixed.

ericpre commented on doc/user_guide/io.rst:
```diff
@@ -1840,28 +1807,46 @@ files automatically.
+  for ``pts`` file only, frames in frame_list will be loaded.
+  for example, ``frame_list=[1,3]`` means second and forth frame will be loaded.
+  If ``None``, all frames are loaded.
+* - ``frame_shifts`` : list of [[int, int]...] or [[int, int, int], ...] or None, default None
```
Can you describe it with more details? For example, when reading it, it is obvious to me why it is possible to pass a "list of lists of length 2" (`[[int, int]...]`) but also a "list of lists of length 3" (`[[int, int, int], ...]`)?
Can you explain to what corresponds the first `int` and the second `int`, etc?

nem1234 replied:
fixed

ericpre reviewed (commented):
Thanks @nem1234, this is in very good shape, I added some small comments/suggestions and there is a test failing on CI (see https://hyperspy.readthedocs.io/en/stable/dev_guide/testing.html#continuous-integration-ci if you want to know more about is meant with CI) because of memory usage:
```python
E           numpy.core._exceptions._ArrayMemoryError: Unable to allocate 15.0 GiB for an array with shape (15, 512, 512, 4096) and data type uint8
```
CI have limited memory, possibly about 2GB, and reduce memory usage of this specific test in one way or another should fix the test failure.

I will need to have another look later at some of the latest changes.

nem1234 committed:
Default values of [] ares replaced with None in pts file reader

nem1234 committed:
Usage comments are updated.
in pts file reader and asw file reader, fd.close before raising exceptions
autoconvert to dense array when compute() lazy array.

nem1234 committed:
Add filename into image title when reading pts file.

nem1234 committed:
Modify pts file title for SEM/STEM image
Modify title check in test_jeol.py

ericpre commented on hyperspy/tests/io/test_jeol.py:
```diff
@@ -245,3 +272,141 @@ def test_shift_jis_encoding():
+    dir2p = dir / 'Sample' / '00_Dummy-Data'
+    file = str(dir2p / test_files2[16])
+    #    dir = TESTS_FILE_PATH
+    #    file = str(dir / 'testpattern1.pts')
```
Is it worth testing these files? If not, they should be removed? If they need to be removed, it would be good to remove them from the history to avoid unnecessarily increasing the size of the git repository. If is a bit annoying to do, I can do ti if necessary.

nem1234 replied:
Ok, I agree to remove those files.
I'm sorry I'm not friendly with complicate GIT operation.
So, please do it on your side.

ericpre commented on hyperspy/io_plugins/jeol.py:
```diff
@@ -453,64 +594,394 @@ def parsejeol(fd):
+        
+    # for lazy loading
+
+    class DenseSliceCOO(sparse.COO):
+        """Just like sparse.COO, but returning a dense array on indexing/slicing"""
+
+        def __getitem__(self, *args, **kwargs):
+            obj = super().__getitem__(*args, **kwargs)
+            try:
+                return obj.todense()
+            except AttributeError:
+                # Indexing, unlike slicing, returns directly the content
+                return obj
```
Instead of copying it here, it is better to import it from `hyperspy/hyperspy/misc/io/fei_stream_readers` to avoid duplicating code.
Ideally, `DenseSliceCOO` should be moved somewhere else because it is generic enough to be used elsewhere than in the velox end reader but we can leave this for another time!

nem1234 replied:
fixed.

ericpre reviewed (commented):
I added some comments: some of them can/should be done in another PR but the one about the test file (comment test code about `testpattern1.pts`) should be sorted in this PR.

nem1234 committed:
Remove duplication code of DenseSliceCOO that comes from hyperspy.misc.io.fei_streamn_readres

nem1234 committed:
Invert direction of frame_shift

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into jeol_pts_emimage

# Conflicts:
#	doc/user_guide/io.rst

ericpre committed:
Add changelog entry and reduce memory usage of the test suite

ericpre commented:
For information I tidy up the history (one merge conflict in `io.rst` wasn't solved correctly and there were unnecessary test files which needed to be removed) by doing the following  
- create a back up by creating a new branch (`jeol_pts_emimage_backup`)
- reset to the head to the commit before the incorrect merge conflict
- cherry pick all commits (4f14a19 to eae6730) after the incorrect merge conflict (from `jeol_pts_emimage_backup` to `jeol_pts_emimage_backup`)
- solve merge conflict in f068f7d
- force push to this branch
- add changelog entry

ericpre commented:
I have reduced the memory usage of some tests, other failures are not related to this PR.

ericpre commented:
Thanks @nem1234, this is a very nice contribution!
